### ISSUE TITLE
SetDigraphVertexLabel List Assignment error

### ISSUE THREAD

**Issue by LRacine** [digraphs/Digraphs#414]
**[gap version using:  4.12dev-638-gcd2de0f]**
**@MTWhyte found that this was due to using DigraphDisjointUnion on mutable digraphs causes D!.vertexlabels to be immutable**
After creating a digraph and calling the SetDigraphVertexLabel function on that digraph, a list assignment error appears.
![image](https://user-images.githubusercontent.com/35891709/108508713-1e1f9c80-72b4-11eb-8c98-c916e83eb148.png)

**Steps for reproduction**
1. load the function:
  ```
CreateTree := function(height)
    local graph, size;
    if (height < 2 ) then
      graph := EmptyDigraph(IsMutable, 1);
      return graph;
    fi;
    
    graph := EmptyDigraph(IsMutable, 1); 
    
    DigraphDisjointUnion(graph, CreateTree(height - 1));
    DigraphAddEdge(graph, 1, 2);
    
    size := Size(DigraphVertices(graph));
    
    DigraphDisjointUnion(graph, CreateTree(height - 2));
    DigraphAddEdge(graph, 1, size + 1);
    
    return graph;
end;
  ```
2. Create a digraph i.e G := CreateTree(5).
3. Call SetDigraphVertexLabel i.e SetDigraphVertexLabel(G, 1, "a");

**Comment by MTWhyte** [comment]
I'm in the process of investigating this, and the problem seems to be that where `D` was created from running `DigraphDisjointUnion` on mutable digraphs, `D!.vertexlabels` is immutable. For example,

```
gap> G := CycleDigraph(IsMutable, 5);
<mutable digraph with 5 vertices, 5 edges>
gap> D := DigraphDisjointUnion(G, G);
<mutable digraph with 10 vertices, 10 edges>
gap> D!.vertexlabels;
[ 1 .. 10 ]
gap> IsMutable(last);
false
```

I haven't fixed it yet, but I thought it would be a good idea to note this on the issue.

**Comment by MTWhyte** [comment]
The trouble seems to be in `DigraphDisjointUnion`. Starting on line `436` of `oper.gi` in its definition, there is the following:

```
  if IsMutableDigraph(D) then
    SetDigraphVertexLabels(D, DigraphVertices(D));
    # This above stops D keeping its old vertex labels after being changed
    ClearDigraphEdgeLabels(D);
    return D;
  fi;
```

The trouble is that `DigraphVertices(D)` returns an immutable list, so ` SetDigraphVertexLabels(D, DigraphVertices(D))` causes `D!.vertexlabels` to be immutable. Setting the vertex labels of a Digraph to be an immutable list will cause the problem, as demonstrated by:

```
gap> A := Immutable([1, 2]);
[ 1, 2 ]
gap> IsMutable(last);
false
gap> D := CycleDigraph(2);
<immutable cycle digraph with 2 vertices>
gap> SetDigraphVertexLabels(D, A);
gap> SetDigraphVertexLabel(D, ["a", "b"]);
Error, no method found! For debugging hints type ?Recovery from NoMethodFound
Error, no 1st choice method found for `SetDigraphVertexLabel' on 2 arguments at /Users/murraywhyte/gap/lib/methsel2.g:249 called from
<function "HANDLE_METHOD_NOT_FOUND">( <arguments> )
 called from read-eval loop at *stdin*:93
type 'quit;' to quit to outer loop
```

There are several ways I can think of to fix this. My first thought we could modify the line `SetDigraphVertexLabels(D, DigraphVertices(D))` to something like `SetDigraphVertexLabels(D, MutableCopyMat(DigraphVertices(D)))` -- or using a more appropriate mutable copy function -- to avoid setting the vertex labels to something immutable during the `DigraphDisjointUnion` process.

A different approach could be modifying `SetDigraphVertexLabels` so that if the list provided is immutable, it converts it to an immutable one. However, it's probably useful in some cases to make the vertex labels immutable, and this would remove that ability.

Finally, `DigraphVertexLabels` could be modified to return a mutable list, instead of an immutable list.

I'm not sure which approach would be most consistent with the package as a whole.

**Comment by LRacine** [comment]
@MTWhyte Thank you for having a look into this and finding the issue! Great to hear this solutions too! For your second solution of adapting SetDigraphVertexLabels, could it return an immutable list if the input digraph had an immutable list and similarly if the input digraph had a mutable list then it would return a mutable list? Would that resolve the issues you had with that solution?

**Comment by james-d-mitchell** [comment]
Thanks @MTWhyte for the detailed explanation. I think the best solution is the one where `SetDigraphVertexLabels` if called with a mutable or immutable digraph and an immutable list of vertex labels, either silently copies the immutable list of vertex labels and makes it mutable, or gives an error. I think the principle should be that the vertex labels of an immutable or mutable  digraph are always mutable, and the best solution is to prevent this from happening in `SetDigraphVertexLabels`.

What say ye @wilfwilson ?

**Comment by wilfwilson** [comment]
I'm not sure actually, I'm going to think about this.

**Comment by wilfwilson** [comment]
@MTWhyte @james-d-mitchell and I just discussed this, and we agreed that the best course of action for now is that, when calling `SetDigraphVertexLabels` with a digraph and an immutable list, then the immutable list is copied with `ShallowCopy` first (which gives an mutable list).

**Comment by LRacine** [comment]
@wilfwilson should this issue be closed?

**Comment by wilfwilson** [comment]
Resolved in v1.4.1.
